File: multi-language/rust/ucd.rs
// multi-language/rust/ucd.rs
struct UCD {
    data: Vec<Vec<f64>>,
}

impl UCD {
    fn new(data: Vec<Vec<f64>>) -> Self {
        UCD { data }
    }

    fn pearson(&self, x: &[f64], y: &[f64]) -> f64 {
        let n = x.len() as f64;
        let mx = x.iter().sum::<f64>() / n;
        let my = y.iter().sum::<f64>() / n;
        let mut num = 0.0;
        let mut dx2 = 0.0;
        let mut dy2 = 0.0;
        for i in 0..x.len() {
            let dx = x[i] - mx;
            let dy = y[i] - my;
            num += dx * dy;
            dx2 += dx * dx;
            dy2 += dy * dy;
        }
        if dx2 * dy2 == 0.0 { 0.0 } else { num / (dx2 * dy2).sqrt() }
    }

    fn analyze(&self, lambda: f64, epsilon: f64) -> (f64, f64, f64, i32) {
        let n = self.data.len();
        if n > 1 {
            let mut sum_corr = 0.0;
            let mut count = 0;
            for i in 0..n {
                for j in i+1..n {
                    sum_corr += self.pearson(&self.data[i], &self.data[j]).abs();
                    count += 1;
                }
            }
            let c = if count > 0 { sum_corr / count as f64 } else { 1.0 };
            let d_eff = n as f64 * (1.0 / (1.0 + lambda));
            let m_size = (10.0 * d_eff / (epsilon * epsilon)).ceil() as i32;
            (c, 1.0 - c, d_eff, m_size)
        } else {
            (0.5, 0.5, 0.0, 0)
        }
    }
}

fn main() {
    let data = vec![
        vec![1.0, 2.0, 3.0, 4.0],
        vec![2.0, 3.0, 4.0, 5.0],
        vec![5.0, 6.0, 7.0, 8.0],
    ];
    let ucd = UCD::new(data);
    let (c, f, d_eff, m_size) = ucd.analyze(0.1, 0.1);
    println!("C: {:.4}, F: {:.4}, d_eff: {:.4}, recommended_m: {}", c, f, d_eff, m_size);
    println!("C: {:.4}, F: {:.4}, d_eff: {:.4}, m: {}", c, f, d_eff, m_size);
}

File: multi-language/cpp/ucd.cpp
// multi-language/cpp/ucd.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <numeric>

class UCD {
    std::vector<std::vector<double>> data;
public:
    UCD(std::vector<std::vector<double>> d) : data(d) {}

    double pearson(const std::vector<double>& x, const std::vector<double>& y) {
        size_t n = x.size();
        double sx = std::accumulate(x.begin(), x.end(), 0.0);
        double sy = std::accumulate(y.begin(), y.end(), 0.0);
        double mx = sx / n, my = sy / n;
        double num = 0, dx2 = 0, dy2 = 0;
        for (size_t i = 0; i < n; ++i) {
            double dx = x[i] - mx;
            double dy = y[i] - my;
            num += dx * dy;
            dx2 += dx * dx;
            dy2 += dy * dy;
        }
        return (dx2 * dy2 == 0) ? 0 : num / std::sqrt(dx2 * dy2);
    }

    std::pair<double, double> analyze() {
        size_t n = data.size();
        if (n > 1) {
            double sumCorr = 0;
            int count = 0;
            for (size_t i = 0; i < n; ++i) {
                for (size_t j = i + 1; j < n; ++j) {
                    sumCorr += std::abs(pearson(data[i], data[j]));
                    count++;
                }
            }
            double c = sumCorr / count;
            return {c, 1.0 - c};
        }
        return {0.5, 0.5};
    }
};

int main() {
    std::vector<std::vector<double>> data = {{1,2,3,4}, {2,3,4,5}, {5,6,7,8}};
    UCD ucd(data);
    auto res = ucd.analyze();
    std::cout << "C: " << res.first << ", F: " << res.second << std::endl;
    return 0;
}
